// @base44/vite-plugin/src/index.ts
import type { Plugin, UserConfig } from "vite";
import { loadEnv } from "vite";
import { errorOverlayPlugin } from "./error-overlay-plugin.js";
import { visualEditPlugin } from "./visual-edit-plugin.js";

const isRunningInSandbox = !!process.env.MODAL_SANDBOX_ID;

export default function vitePlugin(
  opts: {
    legacySDKImports?: boolean;
  } = {}
) {
  const { legacySDKImports = false } = opts;

  return [
    {
      name: "base44",
      config: ({ mode }) => {
        const env = loadEnv(mode ?? "development", process.cwd(), "");

        return {
          resolve: {
            alias: {
              "@/": "/src/",
            },
          },
          ...(isRunningInSandbox
            ? ({
                server: {
                  host: "0.0.0.0", // Bind to all interfaces for container access
                  port: 5173,
                  strictPort: true,
                  // Allow all hosts - essential for Modal tunnel URLs
                  allowedHosts: true,
                  watch: {
                    // Enable polling for better file change detection in containers
                    usePolling: true,
                    interval: 100, // Check every 100ms for responsive HMR
                  },
                  hmr: {
                    protocol: "wss",
                    clientPort: 443,
                  },
                },
                build: {
                  rollupOptions: {
                    onwarn(warning, warn) {
                      // Treat import errors as fatal errors
                      if (
                        warning.code === "UNRESOLVED_IMPORT" ||
                        warning.code === "MISSING_EXPORT"
                      ) {
                        throw new Error(`Build failed: ${warning.message}`);
                      }
                      // Use default for other warnings
                      warn(warning);
                    },
                  },
                },
              } as Partial<UserConfig>)
            : {}),
          optimizeDeps: {
            esbuildOptions: {
              loader: {
                ".js": "jsx",
              },
              ...(legacySDKImports
                ? {
                    define: {
                      "process.env.VITE_BASE44_APP_ID": JSON.stringify(
                        env.VITE_BASE44_APP_ID
                      ),
                      "process.env.VITE_BASE44_BACKEND_URL": JSON.stringify(
                        env.VITE_BASE44_BACKEND_URL
                      ),
                    },
                  }
                : {}),
            },
          },
        };
      },
      resolveId(source, importer, options) {
        if (legacySDKImports) {
          if (source.includes("/entities")) {
            return this.resolve(
              "@base44/vite-plugin/compat/entities.cjs",
              importer,
              options
            );
          }

          if (source.includes("/functions")) {
            return this.resolve(
              "@base44/vite-plugin/compat/functions.cjs",
              importer,
              options
            );
          }

          if (source.includes("/integrations")) {
            return this.resolve(
              "@base44/vite-plugin/compat/integrations.cjs",
              importer,
              options
            );
          }

          if (source.includes("@/agents")) {
            return this.resolve(
              "@base44/vite-plugin/compat/agents.cjs",
              importer,
              options
            );
          }
        }

        return null;
      },
    } as Plugin,
    ...(isRunningInSandbox
      ? [
          {
            name: "iframe-hmr",
            configureServer(server) {
              server.middlewares.use((req, res, next) => {
                // Allow iframe embedding
                res.setHeader("X-Frame-Options", "ALLOWALL");
                res.setHeader("Content-Security-Policy", "frame-ancestors *;");
                next();
              });
            },
          } as Plugin,
          errorOverlayPlugin(),
          visualEditPlugin(),
        ]
      : []),
  ];
}
